src/ConfigProvider.php

```diff
@@ -0,0 +1,137 @@
+                'HttpViewManager' => 'Laminas\Mvc\Service\HttpViewManagerFactory',
+                'InjectTemplateListener' => 'Laminas\Mvc\Service\InjectTemplateListenerFactory',
+                'PaginatorPluginManager' => 'Laminas\Mvc\Service\PaginatorPluginManagerFactory',
+                'Request' => 'Laminas\Mvc\Service\RequestFactory',
+                'Response' => 'Laminas\Mvc\Service\ResponseFactory',
+                'ViewHelperManager' => 'Laminas\Mvc\Service\ViewHelperManagerFactory',
+                View\Http\DefaultRenderingStrategy::class => HttpDefaultRenderingStrategyFactory::class,
+                'ViewFeedStrategy' => 'Laminas\Mvc\Service\ViewFeedStrategyFactory',
+                'ViewJsonStrategy' => 'Laminas\Mvc\Service\ViewJsonStrategyFactory',
+                'ViewManager' => 'Laminas\Mvc\Service\ViewManagerFactory',
+                'ViewResolver' => 'Laminas\Mvc\Service\ViewResolverFactory',
+                'ViewTemplateMapResolver' => 'Laminas\Mvc\Service\ViewTemplateMapResolverFactory',
+                'ViewTemplatePathStack' => 'Laminas\Mvc\Service\ViewTemplatePathStackFactory',
+                'ViewPrefixPathStackResolver' => 'Laminas\Mvc\Service\ViewPrefixPathStackResolverFactory',
+                'Laminas\Mvc\RouteListener' => InvokableFactory::class,
+                'Laminas\Mvc\SendResponseListener' => SendResponseListenerFactory::class,
+                'Laminas\View\Renderer\FeedRenderer' => InvokableFactory::class,
+                'Laminas\View\Renderer\JsonRenderer' => InvokableFactory::class,
+                'Laminas\View\Renderer\PhpRenderer' => ViewPhpRendererFactory::class,
+                'Laminas\View\Strategy\PhpRendererStrategy' => ViewPhpRendererStrategyFactory::class,
+                'Laminas\View\View' => ViewFactory::class,
+
+            ],
+            'shared' => [
+                'EventManager' => false,
+            ],
+            'initializers' => [
+                'EventManagerAwareInitializer' => static function ($first, $second) {
+                    if ($first instanceof ContainerInterface) {
+                        $container = $first;
+                        $instance = $second;
+                    } else {
+                        $container = $second;
+                        $instance = $first;
+                    }
+
+                    if (! $instance instanceof EventManagerAwareInterface) {
+                        return;
+                    }
+
+                    $eventManager = $instance->getEventManager();
+
+                    // If the instance has an EM WITH an SEM composed, do nothing.
+                    if ($eventManager instanceof EventManagerInterface
+                        && $eventManager->getSharedManager() instanceof SharedEventManagerInterface
+                    ) {
+                        return;
+                    }
+
+                    $instance->setEventManager($container->get('EventManager'));
+                },
```

Comment:
We can simplify this by only allowing laminas-servicemanager v3+ and laminas-eventmanager v3+. Personally, I'd do that as part of this patch to simplify this part of the ConfigProvider.

Alternately, move that code into a dedicated method, and use `Closure::fromCallable()` to provide it as an initializer.

Reply:
This change would be outside of the scope for the module manager removal.

I do not see a reason to keep EventManager initializer and expect to drop any initializers entirely as a part of the move towards generic PSR containers. I will need to double check container integrations supported by mezzio, I am not sure if it is something supported by all them.